app/scripts/controllers/metamask-notifications/utils/utils.ts

```diff
@@ -0,0 +1,621 @@
+          k: trigger as TRIGGER_TYPES,
+          e: false,
+        };
+      }
+    }
+  }
+
+  return userStorage;
+}
+
+/**
+ * Upserts (updates or inserts) notification triggers of a specific type across all accounts and chains in user storage.
+ * This method ensures that a trigger of the specified type exists for each account and chain. If a trigger of the specified type
+ * does not exist for an account and chain, it creates a new trigger with a unique UUID.
+ *
+ * @param triggerType - The type of trigger to upsert across all accounts and chains.
+ * @param userStorage - The user storage object to be updated with new or existing triggers of the specified type.
+ * @returns The updated user storage object with upserted triggers of the specified type for all accounts and chains.
+ */
+export function upsertTriggerTypeTriggers(
+  triggerType: TRIGGER_TYPES,
+  userStorage: UserStorage,
+): UserStorage {
+  // Iterate over each account in userStorage
+  Object.entries(userStorage).forEach(([account, chains]) => {
+    if (account === (USER_STORAGE_VERSION_KEY as unknown as string)) {
+      return;
+    }
+
+    // Iterate over each chain for the account
+    Object.entries(chains).forEach(([chain, triggers]) => {
+      // Check if the trigger type exists for the chain
+      const existingTrigger = Object.values(triggers).find(
+        (obj) => obj.k === triggerType,
+      );
+
+      if (!existingTrigger) {
+        // If the trigger type doesn't exist, create a new one with a new UUID
+        const uuid = uuidv4();
+        userStorage[account][chain][uuid] = {
+          k: triggerType,
+          e: false,
+        };
+      }
+    });
+  });
+
+  return userStorage;
+}
+
+/**
+ * Toggles the enabled status of a user storage trigger.
+ *
+ * @param userStorage - The user storage object.
+ * @param address - The user's address.
+ * @param chainId - The chain ID.
+ * @param uuid - The unique identifier for the trigger.
+ * @param enabled - The new enabled status.
+ * @returns The updated user storage object.
+ */
+export function toggleUserStorageTriggerStatus(
+  userStorage: UserStorage,
+  address: string,
+  chainId: string,
+  uuid: string,
+  enabled: boolean,
+): UserStorage {
+  if (userStorage?.[address]?.[chainId]?.[uuid]) {
+    userStorage[address][chainId][uuid].e = enabled;
+  }
+
+  return userStorage;
+}
+
+/**
+ * Attempts to fetch a resource from the network, retrying the request up to a specified number of times
+ * in case of failure, with a delay between attempts.
+ *
+ * @param url - The resource URL.
+ * @param options - The options for the fetch request.
+ * @param retries - Maximum number of retry attempts. Defaults to 3.
+ * @param retryDelay - Delay between retry attempts in milliseconds. Defaults to 1000.
+ * @returns A Promise resolving to the Response object.
+ * @throws Will throw an error if the request fails after the specified number of retries.
+ */
+async function fetchWithRetry(
```

Comment:
See prev comment on retries. IDK if this should live in controller level or on UI level.

It feels weird to re-invent the wheel here when RTK-Query or other data/state management libs can handle this.

Lets refactor this in a future PR once we connect and play with the UI.